common/Storyboarding/OsbSpriteWriter.cs

```diff
@@ -0,0 +1,306 @@
+
+            return !(moveTimeline.HasOverlap ||
+                     moveXTimeline.HasOverlap ||
+                     moveYTimeline.HasOverlap ||
+                     rotateTimeline.HasOverlap ||
+                     scaleTimeline.HasOverlap ||
+                     scaleVecTimeline.HasOverlap ||
+                     fadeTimeline.HasOverlap ||
+                     colorTimeline.HasOverlap);
+        }
+
+        protected virtual HashSet<int> GetFragmentationTimes()
+        {
+            HashSet<int> fragmentationTimes = new HashSet<int>();
+            var nonFragmentableCommands = OsbSprite.Commands.Where(c => !c.IsFragmentable()).ToList();
+
+            fragmentationTimes.UnionWith(Enumerable.Range((int)OsbSprite.StartTime, (int)(OsbSprite.EndTime - OsbSprite.StartTime)));
+            
+            nonFragmentableCommands.ForEach(c =>
+            {
+                var range = Enumerable.Range((int)c.StartTime + 1, (int)(c.EndTime - c.StartTime - 1));
+                fragmentationTimes.ExceptWith(range);
+            });
+
+            return fragmentationTimes;
+        }
+
+        private List<ICommand> GetNextSegment(HashSet<int> fragmentationTimes, List<ICommand> commands)
+        {
+            List<ICommand> segment = new List<ICommand>();
+
+            int startTime = fragmentationTimes.Min();
+            int endTime;
+            int maxCommandCount = OsbSprite.MaxCommandCount;
+
+            //split the last 2 segments evenly so we don't have weird 5 command leftovers
+            if (commands.Count < OsbSprite.MaxCommandCount * 2 && commands.Count > OsbSprite.MaxCommandCount)
+                maxCommandCount = (int)Math.Ceiling(commands.Count / 2.0);
+
+            if (commands.Count < maxCommandCount)
+                endTime = fragmentationTimes.Max() + 1;
+            else
+            {
+                var cEndTime = (int)commands.OrderBy(c => c.StartTime).ElementAt(maxCommandCount - 1).StartTime;
+                if (fragmentationTimes.Contains(cEndTime))
+                    endTime = cEndTime;
+                else
+                {
+                    endTime = fragmentationTimes.Where(t => t < cEndTime).Max();
+                    if (endTime == startTime) // segment can't be <= MaxCommandCount, so we use the smallest available
+                        endTime = fragmentationTimes.First(t => t > startTime);
+                }
+            }
+
+            foreach (var cmd in commands.Where(c => c.StartTime < endTime))
+            {
+                var sTime = Math.Max(startTime, (int)Math.Round(cmd.StartTime));
+                var eTime = Math.Min(endTime, (int)Math.Round(cmd.EndTime));
+                ICommand command;
+                if (sTime == (int)Math.Round(cmd.StartTime) && eTime == (int)Math.Round(cmd.EndTime))
+                {
+                    command = cmd;
+                }
+                else
+                {
+                    var type = cmd.GetType();

```

Comment:
You're splitting the command here? add a split method to commands instead of this magic

Reply:
done